src/wp-admin/includes/misc.php

```diff
@@ -1641,3 +1641,133 @@ function wp_check_php_version() {
 
 	return $response;
 }
+
+/**
+ * Creates and returns the markup for an admin notice.
+ *
+ * @since 6.3.0
+ *
+ * @param string $message The message.
+ * @param array  $args {
+ *     Optional. An array of arguments for the admin notice. Default empty array.
+ *
+ *     @type string   $type               Optional. The type of admin notice.
+ *                                        For example, 'error', 'success', 'warning', 'info'.
+ *                                        Default empty string.
+ *     @type bool     $dismissible        Optional. Whether the admin notice is dismissible. Default false.
+ *     @type string   $id                 Optional. The value of the admin notice's ID attribute. Default empty string.
+ *     @type string[] $additional_classes Optional. A string array of class names. Default empty array.
+ *     @type bool     $paragraph_wrap     Optional. Whether to wrap the message in paragraph tags. Default true.
+ * }
+ */
+function wp_get_admin_notice( $message, $args ) {
+	$defaults = array(
+		'type'               => '',
+		'dismissible'        => false,
+		'id'                 => '',
+		'additional_classes' => array(),
+		'paragraph_wrap'     => true,
+	);
+
+	$args = wp_parse_args( $args, $defaults );
+
+	/**
+	 * Filters the arguments for an admin notice.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param array  $args    The arguments for the admin notice.
+	 * @param string $message The message for the admin notice.
+	 */
+	$args    = apply_filters( 'wp_admin_notice_args', $args, $message );
+	$id      = '';
+	$classes = 'notice';
+
+	if ( is_string( $args['id'] ) ) {
+		$trimmed_id = trim( $args['id'] );
+
+		if ( '' !== $trimmed_id ) {
+			$id = 'id="' . esc_attr( $trimmed_id ) . '" ';
+		}
+	}
+
+	if ( is_string( $args['type'] ) ) {
+		$type = trim( $args['type'] );
+
+		if ( '' !== $type ) {
+			if ( 'updated' === $type ) {
+				$type = 'success';
+			}
+
+			if (
+				// Don't add "notice-" when it already exists.
+				str_starts_with( $type, 'notice-' ) ||
+				// Don't add "notice-" for a type containing spaces.
+				str_contains( $type, ' ' ) ||
+				// Don't add "notice-" for unknown types.
+				! in_array( $type, array( 'error', 'success', 'warning', 'info' ), true )
+			) {
+				$classes .= ' ' . $type;
+			} else {
+				$classes .= ' notice-' . $type;
+			}
+		}
+	}
+
+	if ( true === $args['dismissible'] ) {
+		$classes .= ' is-dismissible';
+	}
+
+	if ( is_array( $args['additional_classes'] ) && ! empty( $args['additional_classes'] ) ) {
+		$classes .= ' ' . implode( ' ', $args['additional_classes'] );
+	}
+
+	if ( false !== $args['paragraph_wrap'] ) {
+		$message = "<p>$message</p>";
```

Comment:
What escaping (if any) does `$message` need?

Reply:
I think none here; escaping should be done at echo, and this function never echoes the value.

Reply:
Thanks Joe! On that note, the attributes are escaped in this function, as it would otherwise need to escape the whole notice markup. What are your thoughts on this?

Reply:
Thanks Joe! On that note, the attributes are escaped in this function, otherwise the whole notice markup would need to be escaped before outputting. What are your thoughts on this? If we don't escape anything in this function, what's a sufficient escaping function when outputting in `wp_admin_notice()` that'll handle the attributes too?

Reply:
`wp_kses_post()` is pretty efficient on small quantities of content, so that would probably be fine. We use it in the settings API, which is fairly comparable in terms of how it can be used by extenders, so it seems reasonable.

Reply:
- Escaping removed from `wp_get_admin_notice()` in [6c16346](https://github.com/WordPress/wordpress-develop/pull/4119/commits/6c16346b9cc62e9716fbae823a9fef0e21f1abf0).
- Escaping added to `wp_admin_notice()` in [efe75d5](https://github.com/WordPress/wordpress-develop/pull/4119/commits/efe75d52fe1e73825baff1b55b5dda43642b3cea).